buildpacks/gradle/src/gradle_command/tasks.rs

```diff
@@ -0,0 +1,189 @@
+use crate::gradle_command::{run_gradle_command, GradleCommandError};
+use libcnb::Env;
+use std::path::Path;
+use std::process::Command;
+
+pub(crate) fn tasks(
+    current_dir: &Path,
+    env: &Env,
+) -> Result<Tasks, GradleCommandError<nom::error::Error<String>>> {
+    run_gradle_command(
+        Command::new(current_dir.join("gradlew"))
+            .current_dir(current_dir)
+            .envs(env)
+            .args(["--quiet", "tasks"]),
+        |stdout, _stderr| {
+            parser::parse(stdout)
+                .map(|groups| Tasks { groups })
+                .map_err(|error| nom::error::Error {
+                    input: error.input.to_string(),
+                    code: error.code,
+                })
+        },
+    )
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub(crate) struct Tasks {
+    pub(crate) groups: Vec<TaskGroup>,
+}
+
+impl Tasks {
+    pub(crate) fn names(&self) -> Vec<String> {
+        self.groups
+            .iter()
+            .flat_map(|task_group| &task_group.tasks)
+            .map(|task| task.name.clone())
+            .collect()
+    }
+
+    pub(crate) fn has_task(&self, s: &str) -> bool {
+        self.names().iter().any(|task_name| task_name == s)
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub(crate) struct TaskGroup {
+    pub(crate) heading: String,
+    pub(crate) tasks: Vec<Task>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub(crate) struct Task {
+    pub(crate) name: String,
+    pub(crate) description: String,
+}
+
+mod parser {
+    use super::Task;
+    use super::TaskGroup;
+    use nom::bytes::complete::tag;
+    use nom::character::complete::{alphanumeric1, char, line_ending, not_line_ending};
+    use nom::combinator::{map, verify};
+    use nom::multi::{count, many0, many1, many_till};
+    use nom::sequence::{terminated, tuple};
+    use nom::{Finish, IResult};
+
+    pub(crate) fn parse(input: &str) -> Result<Vec<TaskGroup>, nom::error::Error<&str>> {
+        many1(map(many_till(any_line, task_group), |(_, out)| out))(input)
+            .finish()
+            .map(|(_remaining, parsed)| parsed)
+    }
+
+    fn task_group(input: &str) -> IResult<&str, TaskGroup> {
+        map(
+            tuple((task_group_heading, many0(task_line))),
+            |(heading, lines)| TaskGroup {
+                heading,
+                tasks: lines,
+            },
+        )(input)
+    }
+
+    fn task_group_heading(input: &str) -> IResult<&str, String> {
+        let (input, line) = verify(any_line, |line: &str| line.ends_with("tasks"))(input)?;
+        let (input, _) = terminated(count(char('-'), line.len()), line_ending)(input)?;
+
+        Ok((input, line))
+    }
+
+    fn task_line(input: &str) -> IResult<&str, Task> {
+        map(
+            tuple((
+                map(terminated(alphanumeric1, tag(" - ")), String::from),
```

Comment:
can a task name start with a number?  also, is ASCII too restrictive?  i believe gradle defines their disallowed characters [here](
https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/util/internal/NameValidator.java#L26-L27).

Reply:
Yeah, they technically can contain more than `alphanumeric1` allows. 🤔 Thanks for pointing this out - I will adjust.